RecoTracker/ExoRegionalTracking/plugins/ExoRegionProducer.cc

```diff
@@ -0,0 +1,231 @@
+  ExoRegionTFCache *cache = new ExoRegionTFCache();
+  cache->graphDef = tensorflow::loadGraphDef(pbFile);
+
+  return unique_ptr<ExoRegionTFCache>(cache);
+}
+
+void ExoRegionProducer::globalEndJob(const ExoRegionTFCache *cache) {
+  if (cache->graphDef != nullptr)
+    delete cache->graphDef;
+}
+
+void ExoRegionProducer::produce(edm::Event &event, const edm::EventSetup &setup) {
+  edm::Handle<reco::BeamSpot> beamSpot;
+  edm::Handle<edm::View<reco::VertexCompositeCandidate> > trackClusters;
+  edm::Handle<edm::View<reco::GenParticle> > genParticles;
+
+  event.getByToken(beamSpotToken_, beamSpot);
+  const math::XYZVector bs(beamSpot->x0(), beamSpot->y0(), beamSpot->z0());
+  event.getByToken(trackClustersToken_, trackClusters);
+  vector<math::XYZVector> decayVertices;
+  if (genMatch_ != "None") {
+    event.getByToken(genParticlesToken_, genParticles);
+    getDecayVertices(*genParticles, decayVertices);
+  }
+
+  // Initialize distances.
+  list<ExoRegion> pseudoROIs;
+  list<Distance> distances;
+  const double minTrackClusterRadius = minRadius_ - rParam_;
+  for (unsigned i = 0; i < trackClusters->size(); i++) {
+    const reco::VertexCompositeCandidate &trackCluster = trackClusters->at(i);
+    const math::XYZVector x(trackCluster.vx(), trackCluster.vy(), trackCluster.vz());
+    if (minRadius_ < 0.0 || minTrackClusterRadius < 0.0 || (x - bs).rho() > minTrackClusterRadius)
+      pseudoROIs.emplace_back(trackClusters, i, rParam_);
+  }
+  if (pseudoROIs.size() > 1) {
+    ExoRegionItr secondToLast = pseudoROIs.end();
+    secondToLast--;
+    for (ExoRegionItr i = pseudoROIs.begin(); i != secondToLast; i++) {
+      ExoRegionItr j = i;
+      j++;
+      for (; j != pseudoROIs.end(); j++)
+        distances.emplace_back(i, j);
+    }
+  }
+
+  // Track clusters farther apart than 4 times rParam_ cannot wind up in the
+  // same ROI, so remove these pairs.
+  const auto pred = [&](const Distance &a) { return a.distance() > 4.0 * rParam_; };
+  distances.remove_if(pred);
+
+  // Do clustering.
+  while (!distances.empty()) {
+    const auto comp = [](const Distance &a, const Distance &b) { return a.distance() <= b.distance(); };
+    distances.sort(comp);
+    DistanceItr d = distances.begin();
+    if (d->distance() > rParam_)
+      break;
+
+    d->entities().first->merge(*d->entities().second);
+    d->entities().second->setInvalid();
+
+    const auto distancePred = [](const Distance &a) {
+      return (!a.entities().first->valid() || !a.entities().second->valid());
+    };
+    const auto pseudoROIPred = [](const ExoRegion &a) { return !a.valid(); };
+    distances.remove_if(distancePred);
+    pseudoROIs.remove_if(pseudoROIPred);
+  }
+
+  // Remove invalid ROIs.
+  const auto roiPred = [&](const ExoRegion &roi) {
+    if (!roi.valid())
+      return true;
+    const math::XYZVector x(roi.vx(), roi.vy(), roi.vz());
+    if ((minRadius_ >= 0.0 && (x - bs).rho() < minRadius_) ||
```

Comment:
do you really need the check of being a positive value ?

Reply:
You're right! The check has been removed.